### ISSUE TITLE
[Bug]: Rendering order is different on initial fetch and reload fetch using a loader with defer

### ISSUE THREAD

**Issue by AugustinUpway** [remix-run/react-router#11228]
### What version of React Router are you using?

6.21.3

#### Context

My team is building a filtered collection page for a Shopify Store. 
It will be a client rendered widget at first but we're building it with React Router for easily switching to Remix later !

#### Use case

We have two queries, one to get a product list and another to get a list of available filter for those products.
The latter is way longer than the former.
We'd like to have data loaded in parallel to the route and maybe waiting for the faster request before showing anything (aka what's defer is made for).

Last but not least, you can apply filter on this route (aka click a link that add search params to the url) that refresh the page data. On those refresh, we'd like to see each component as soon as the updated data comes in.

Code is fairly simple:
```ts
const Rootloader = (async () => {
  const productsPromise = fetchProducts();
  const filtersPromise = fetchFilters();
  return defer({
    products: productsPromise, // Fast query, could be awaited at some point
    filters: filtersPromise, // Slower query
  });
}) satisfies LoaderFunction;

const Root: FC = () => {
  const data = useLoaderData<typeof Rootloader>();
  return (
    <div>
      <Suspense fallback={'Filters are loading'}>
        <Await resolve={data.filters}>{(f) => <Filters filters={f} />}</Await>
      </Suspense>
      <Suspense fallback={'Products are loading'}>
        <Await resolve={data.products}>{(p) => <Products products={p} />}</Await>
      </Suspense>
    </div>
  );
};
```

[Full example I've used here (gist)](https://gist.github.com/AugustinUpway/cbce589a872aea0416a2acd85751ccdb)

### Expected Behavior

Rendering timeline should look the same on data refresh than on initial fetch.

In my exemple I'd like to see:
- product fetch done -> product component updated
- filter fetch done -> filter component updated

### Actual Behavior

Rendering timeline is different on data refresh than on initial fetch.

On initial fetch we got:
- product fetch done -> product component updated
- filter fetch done -> filter component updated

But on data refresh, we end but with:
- product fetch done -> wait for filter fetch -> product component updated
- filter fetch done -> wait for product fetch -> filter component updated

If I resume this with an annotated log drawing:

![image](https://github.com/remix-run/react-router/assets/106661384/70a70bea-ae8d-4f97-bd99-dff619f58488)

The promise become interlocked on data refresh.
Is this a bug ? A missing feature ?
Is there something we don't know that would make this work ?
Is defer not made for this at all ?

Happy to provide more information if needed !
Also I'm not proficient with React Router code base but could make a doc PR if this could help others understand what going on 😃 

Thanks for your awesome work !

**Comment by AugustinUpway** [comment]
Hi @brophdawg11, thx for your answer !
We removed the `defer` to move on as suggested.

However, sorry to insist, in our case: we don't want to get the suspense fallbacks on revalidations.

Instead we'd like to have the most up to date values coming from from our resolved deferred promises, mixed with stale deffered values for the remaining ongoing promises.
But React Router does not do that, it does one render with all stale data and one with all up to date data.

Feel like using defer would have been fine on our case if our loader was working with only one promise, but because we have two, it does not make the cut.

Are we talking about the same thing ?
I'm definitely not an expert on React Suspense 😅

**Comment by AugustinUpway** [comment]
I've built another example where the data fetching is done with two different loader but the behavior on revalidation is the same. Hope this help ! 🤞 

```typescript
// The data fetching is now split between two loaders

const Stage1Loader = (async () => {
  const productsPromise = fetchProducts();
  return defer({ products: productsPromise, });
}) satisfies LoaderFunction;

const Stage2Loader = (async () => {
  const filtersPromise = fetchFilters();
  return defer({ filters: filtersPromise, });
}) satisfies LoaderFunction;

// Part of the UI is now nested into the other

const Stage1: FC = () => {
  const data = useLoaderData<typeof Stage1Loader>();
  return (
    <div>
      <Outlet />
      <div>
        <h3>Products</h3>
        <Suspense fallback={'Products are loading (stage 1)'}>
          <Await resolve={data.products}>{(p) => <Products products={p} />}</Await>
        </Suspense>
      </div>
    </div>
  );
};

const Stage2: FC = () => {
  const data = useLoaderData<typeof Stage2Loader>();
  return (
    <div>
      <h3>Filters</h3>
      <Suspense fallback={'Filters are loading (stage 2)'}>
        <Await resolve={data.filters}>{(f) => <Filters filters={f} />}</Await>
      </Suspense>
    </div>
  );
};

const router = createHashRouter([
  {
    path: '*',
    element: <Navigate to={'/stage1/stage2'} replace />,
  },
  {
    path: '/stage1',
    loader: Stage1Loader, // On revalidation, we'll not see this data until the nested loader is done, why ?
    element: <Stage1 />,
    children: [
      {
        path: '/stage1/stage2',
        loader: Stage2Loader,
        element: <Stage2 />,
      },
    ],
  },
]);
```

**Comment by brophdawg11** [comment]
Ah, ok thanks for the explanation.  React Router doesn't do any caching of stale values - it only ever knows about the latest loader return value per route.  So as soon as you return a new promise, the old resolved value is gone.  You need to manage your own cache of that stale data if you'd like to render it while waiting on the fresh data.  It's basically an SWR approach.

Here's an example based on how this is done in [`remix-client-cache`](https://github.com/Code-Forge-Net/remix-client-cache) which does this in Remix via `clientLoader`'s:

https://codesandbox.io/p/sandbox/naughty-yalow-l8czh7?file=%2Fsrc%2Findex.js%3A125%2C1
